powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/totp/TOTP.java

```diff
@@ -0,0 +1,299 @@
+        return validateTOTP(otp, key, localDateTime, steps, Algorithm.HMAC_SHA256.code);
+    }
+
+    /**
+     * Validate a TOTP value for the given set of parameters using HmacSHA512 algorithm. Validates one time step backward.
+     *
+     * @param otp           TOTP to validate
+     * @param key           the shared secret
+     * @param localDateTime date time
+     * @return true if OTP is valid
+     * @throws CryptoProviderException in case of any crypto error
+     * @see #validateTOTPSHA1(String, byte[], LocalDateTime, int)
+     */
+    public static boolean validateTOTPSHA512(final String otp, final byte[] key, final LocalDateTime localDateTime) throws CryptoProviderException {
+        return validateTOTPSHA512(otp, key, localDateTime, 1);
+    }
+
+    /**
+     * Validate a TOTP value for the given set of parameters using HmacSHA512 algorithm.
+     *
+     * @param otp           TOTP to validate
+     * @param key           the shared secret
+     * @param localDateTime date time
+     * @param steps         number of backward time steps allowed to validate
+     * @return true if OTP is valid
+     * @throws CryptoProviderException in case of any crypto error
+     */
+    public static boolean validateTOTPSHA512(final String otp, final byte[] key, final LocalDateTime localDateTime, final int steps) throws CryptoProviderException {
+        return validateTOTP(otp, key, localDateTime, steps, Algorithm.HMAC_SHA512.code);
+    }
+
+    /**
+     * Validate a TOTP value for the given set of parameters using HmacSHA1 algorithm.
+     *
+     * @param otp           TOTP to validate
+     * @param key           the shared secret
+     * @param localDateTime date time
+     * @param steps         number of backward time steps allowed to validate
+     * @param algorithm     the algorithm to use
+     * @return true if OTP is valid
+     * @throws CryptoProviderException in case of any crypto error
+     */
+    private static boolean validateTOTP(final String otp, final byte[] key, final LocalDateTime localDateTime, final int steps, final String algorithm) throws CryptoProviderException {
+        logger.debug("Validating TOTP for localDateTime={}, algorithm={}, steps={}", localDateTime, algorithm, steps);
+
+        if (otp == null || otp.trim().length() == 0) {
+            throw new CryptoProviderException("Otp is mandatory");
+        }
+
+        if (steps < 0) {
+            throw new CryptoProviderException("Steps must not be negative number");
+        }
+
+        long currentTimeStep = countTimeSteps(localDateTime);
+        for (int i = 0; i <= steps; i++) {
+            logger.debug("Validating TOTP for localDateTime={}, algorithm={}, backward step={} out of allowed steps={}", localDateTime, algorithm, i, steps);
+            final long step = currentTimeStep - i;
+            final String expectedOtp = generateTOTP(key, step, otp.length(), algorithm);
+            if (expectedOtp.equals(otp)) {
```

Comment:
This section is prone to side channel timing attacks, although the TOTP is really short, we should probable still avoid code branches. We should compare all parts of the TOTP everytime.